### DISCUSSION TITLE
LDM Notes for October 2nd, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8499] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-10-02.md

### Agenda

- Open questions in field
    - `readonly` contexts and `set`
    - `Conditional` code
    - Interface properties and auto-accessors
- Extensions

**Comment by DavidArno** [comment]
Regarding:
> ...One possibility is a single pass, that pretends the type parameters from the extension type are part of the method signature. Then we do the same single pass as in extension methods today, and it all behaves the same.
>
> This comes with its own downsides, however, which is breaking a different expectation from C# users around how type parameters work...

One possible approach here is to adopt the two stage approach (it fits with expectations on how type parameters work today), but introduce some syntax that implies "_re-run stage 1 here_", eg:
```cs
public static extension E2<T> for I<T>
{
   public static void M2<T2>(out T2 t) where T2 is T { t = default; }
}
```
`is` is used here simply to be a different thing to `:`. The exact syntax isn't important. It instructs the compiler to  re-evaluate `T` within the context of the method. This would allow for the two stage approach, would maintain consistency with existing behaviour and would solve this use case.

**Comment by TahirAhmadov** [comment]
Regarding LINQ, I can't see a way to keep it binary compatible. The way forward would be designing new version with the 2 stage process in mind, marking existing extension methods obsolete, and using overload resolution system to prefer the new extensions (albeit I'm not sure right now exactly how to achieve a smooth transition regarding overload resolution - is it the overload priority attribute? )

**Reply by MgSam** [reply]
It sounds like the conclusion LDT reached here was that `extension` types themselves are unworkable in a way to maintain backwards compat. Now that the "typeness" of `extension` has been removed I personally think there's little value in continuing to give it a whole new type syntax anyway.

Perhaps just extend the a new syntax form for properties- that's 99% of what people wanted anyway.

Strawman:
``` C#
public static class MyExtensions 
{
    //Extension property 
    public static string ExtensionProp for string => this.Length;

    //Traditional extension method
    public static string MyLength(this string s) => s.Length;
}
```
I'm not tied to any specific syntax for extension property- just suggesting the LDT go back to this approach.

**Reply by TahirAhmadov** [reply]
> Perhaps just extend the a new syntax form for properties- that's 99% of what people wanted anyway.

I don't think that's the case, though. Many also want extension static members, and extension interface implementation is still a future goal.

At this point, given that all of this discussion hinges on runtime making the zero-cost struct wrapper work reliably, which will give us everything without the need for implementation changes, I'd say (I know, people will hate me for this) - just postpone shipping extensions until such runtime improvements are made.

**Reply by HaloFour** [reply]
I recall generic indexers being one of the use cases that felt... unwieldy when the syntax for extension properties was being tossed around:

```cs
public static string this<T>[int index] for T where T : IFoo {
    get { ... }
}
```

It would probably make the design space around extension operators similarly complicated, especially generic ones.

```cs
// bad example, just for illustration of syntax
public static TResult operator+<T, TOther, TResult>(T x, TOther y) for T
    where T : IAdditionOperators<T, TOther, TResult> {

    return x + y;
}
```

I think these are very important thought experiments, though.  How important is it for extension methods to be able to "grow up" into extensions?  Is this feature intended to fully replace extension methods?  I personally think there is room for both "extension methods" and "extension types", and the concern over how LINQ might have been implemented under this new design highlights the reason why.

**Reply by TahirAhmadov** [reply]
I always thought extension methods would be de facto (probably not de jure) obsoleted. What's the point of keeping 2 approaches at the same time?

From the meeting notes, it sounds like the LINQ concerns are related to the inability of making existing LINQ extension methods be binary compatible with new extensions feature. However, I don't really understand why that is a goal to begin with. In fact, given that at some point in the near future the runtime will be improved to allow "the real type" approach, which also allows additional interface implementation, that version cannot be binary compatible with existing extension methods at all.

The only way forward to stay binary compatible with extension methods is to have a dual approach: a `static class` to hold the members (with all of the transpilation involved with that), and a separate struct for interface implementation; to me, this sounds like way more trouble than it's worth.

And WRT abolishing extensions altogether and simply adding more member types under the existing extension method paradigm, what about interface implementation? Either somebody has to come up with a completely new syntax for that, or interface implementation goal is given up on, or this approach isn't workable at all.

**Reply by TahirAhmadov** [reply]
WRT binary compatibility, the forthcoming code fixer should have the option to "convert-and-delete" existing extension methods, and the option to "convert-and-retain". This way, binary compatibility can be preserved in those cases where it's needed (BCL, some other libraries/Nugets/etc.). Whereas in a lot of "end-user" (not reusable) code the "delete" option is used and everybody is happy.

**Reply by dersia** [reply]
I agree, why is binary compact a goal? why not keep the old way of writing extensions methods for let's say 2 versions, but obsolted and then drop support for it and only support the new way. this might be done source compatible.

**Reply by CyrusNajmabadi** [reply]
> I agree, why is binary compact a goal?

So we don't break realistically 100% of all projects out there. 

> and only support the new way. 

This breaks a massive amount of projects.

**Reply by CyrusNajmabadi** [reply]
> What's the point of keeping 2 approaches at the same time?

Compat.  To be clear, we discuss with representatives who own libraries used by *the majority* of users who aren't even willing to accept a new way of doing extensions if they're not binary compatible.  And even if they are binary compatible, it's a very tough sell to still update their form. It's an extremely risky thing for them to do no matter what. At least with perfect binary compat the can look at the assembly before/after and see that nothing changed and that this is safe. 

This is the same path we took with nrt which had equal ecosystem need to not break people who had been doing things in a certain way for 20+ years. 

Remember, that's how old extensions are. Think about all the code since then that created and consumed those extensions over virtually every combination of runtime, dll, compilers, etc. 

You may be on a golden path. Or your risk averseness might be lower than others. But our designs need to encompass major API authors that must be much more careful and who end up wanting compat for *their* millions of customers.

**Reply by CyrusNajmabadi** [reply]
> are related to the inability of making existing LINQ extension methods be binary compatible with new extensions feature.

No. It's about making *any* new extension method binary compatible. This is not limited to linq.

**Reply by jnm2** [reply]
> So we don't break realistically 100% of all projects out there.

This would only happen if library authors threw caution to the wind; alternatively, they'd keep their classic extension methods as classic extension methods. Library authors that I've worked with are suspicious of switching to new language features that affect ABI, and they're also intimately familiar with leaving old APIs around for compat.

**Reply by TahirAhmadov** [reply]
> Compat. To be clear, we discuss with representatives who own libraries used by _the majority_ of users who aren't even willing to accept a new way of doing extensions if they're not binary compatible. And even if they are binary compatible, it's a very tough sell to still update their form. It's an extremely risky thing for them to do no matter what. At least with perfect binary compat the can look at the assembly before/after and see that nothing changed and that this is safe.

That is solved perfectly by the "convert and retain" code fixer idea of mine. 

In terms of DRY, the remaining extension methods get refactored to call the new extension types' methods, voila.

**Reply by En3Tho** [reply]
Is the goal is not to change/break existing extensions codegen while doing new one? 

Seems reasonable.

Is the goal is to somehow tie new extensions codegen to the old one because of risk some million-user-library author might change the code thinking there will be no difference? And somehow that new versioned dll will leak into the production app bypassing all build/test/ci process?

Feels overly protective tbh. I'm sure these people can comprehend the idea of "you know, these are actually generated differently".

**Reply by CyrusNajmabadi** [reply]
> > Compat. To be clear, we discuss with representatives who own libraries used by _the majority_ of users who aren't even willing to accept a new way of doing extensions if they're not binary compatible. And even if they are binary compatible, it's a very tough sell to still update their form. It's an extremely risky thing for them to do no matter what. At least with perfect binary compat the can look at the assembly before/after and see that nothing changed and that this is safe.
> 
> That is solved perfectly by the "convert and retain" code fixer idea of mine. 
> 
> In terms of DRY, the remaining extension methods get refactored to call the new extension types' methods, voila.

The question was "why is binary compare a goal". I was answering why.

**Reply by CyrusNajmabadi** [reply]
> Library authors that I've worked with are suspicious of switching to new language features that affect ABI,

Right. Which is why one of the proposals suggests not changing the ABI.  To avoid even having to consider that.

**Reply by TahirAhmadov** [reply]
> The question was "why is binary compare a goal". I was answering why.

I understand the need to preserve binary compatibility at the library level, my point was the extension binary compat is not needed to achieve that goal - just keeping existing code in place solves that very cheaply.

**Reply by HaloFour** [reply]
I think having a safe way to evolve from one syntax to the other would be beneficial for the language as whole, otherwise we will always be in this state where some extensions live in one form and other extensions live in another.  It would otherwise probably dissuade libraries from adopting the new syntax, as they won't accept the breaking change and don't want to exist in a split-brain model.

**Reply by CyrusNajmabadi** [reply]
> just keeping existing code in place solves that very cheaply.

That is a form of back compat. I was answering someone who was arguing we didn't need that, and that we could deprecate and remove the existing code approach.

**Reply by jnm2** [reply]
> otherwise we will always be in this state where some extensions live in one form and other extensions live in another.

How could we not be in this state regardless? C# is a language that allows multiple ways to do the same thing, and many codebases choose to stick with the familiar as new features come out. I think this is a great thing about C#.

**Reply by CyrusNajmabadi** [reply]
> How could we not be in this state regardless? 

There's a difference between "i only can be in this state" and "i can choose to be in this state".

I'm saying: if we don't have compat in the new form, your only choice is keeping your own extnesions around.

**Reply by HaloFour** [reply]
Exactly.  Imagine if there were cases in the language where, due to binary compatibility reasons, you were _required_ to use the anonymous delegate syntax instead of lambdas.  :)

**Reply by CyrusNajmabadi** [reply]
Yup.  It's something we *could* do.  But views differ on how palatable that is.  To me, it's not particularly palatable.  I dont' really want to say: here are your legacy extensions, you can never touch them, and they always have to stand apart from the other extensions you add with the new system.  Yes, that's a *technically possible* path.  It's just not very attractive.  I wnt to be able to customers who *do* want a consistent codebase: yes, you can have that.  You can use new extensions without a single concern about compat moving your classic extensions to that form.

**Reply by jnm2** [reply]
> Exactly. Imagine if there were cases in the language where, due to binary compatibility reasons, you were required to use the anonymous delegate syntax instead of lambdas. :)

This is stated like it would obviously be a bad thing, but I can imagine that. Library authors can't use every new C# feature within public APIs, or have to handle them with great care. Library authors are also familiar with leaving old APIs around when introducing new ones. I'm open-minded on how much of a priority it would be to provide the option to move everything with perfect compat.

I'm not against it, but there are two things:
- It can be done later at any time, after seeing how things go.
- It's not as though this is a new situation for library authors to be in. Remember when NRTs came out, and the official guidance (initially) was not to annotate anything that targeted netstandard or netframework, which realistically almost all libraries did multi-target against in addition to .NET Core?

**Reply by CyrusNajmabadi** [reply]
> This is stated like it would obviously be a bad thing, but I can imagine that. Library

It would personally be a bad thing for me if new features were things that users could not move forward onto even if they want to.  That's different from saying that all users need to move forward.  My preference is a world where people have the choice.   If they don't care, or actually prefer to stick with the existing syntax, that's 100% fine.  But if they would prefer to move forward they should have that option.  That's very different then from giving them something new and saying they *cannot* move forward.

Again, it's about a choice.  If there is a choice, both groups can have what they want.  Group A (which doesn't want to move) doesn't have to.  Group B (which does want to move) can.  

If we force users to move, that hurts group A.  If we make it so that you cannot move safely, that hurts group B.  I would prefer both groups have their needs met.

**Reply by dersia** [reply]
maybe I communicated my previous point wrongly, I am sorry for that. my understanding was that it seems to be very hard to keep both solutions and therefor I sad, why would we need to have a binary compat hanging on forever with, in this case I would prefer dropping the binary compat at some point or at least deprecated it and stop evolving it.

that being said, I would also love to support both groups, but I would hate it to drop the new planned extensions feature just to keep the group happy that "only" gets extions method, if they could just move forward to the new way of extensions.

I hope this clarified my point. I think the new extensions feature and what it can become is worth so much more than just keeping the old "extensions methods only" feature

**Reply by TahirAhmadov** [reply]
> My preference is a world where people have the choice.

They will have the choice, though. I'm not sure what the concern is here.
1. If they want to keep using extension methods, fine. 
2. If they want to ditch them altogether and transition to extension types, fine.
3. If they want to offer extension types but also keep the extension methods for compatibility, that's also possible:
```cs
extension StringExt for string
{
  public int NumLetters => this.Where(c=>char.IsLetter(c)).Count();
  public int GetNumDigits() => this.Where(c=>char.IsDigit(c)).Count();
}
static class StringExtensions
{
  // used to be
  public static int GetNumLetters(this string This) => This.Where(c=>char.IsLetter(c)).Count();
  public static int GetNumDigits(this string This) => This.Where(c=>char.IsDigit(c)).Count();

  // now is
  [EditorBrowsable(EditorBrowsableState.Never)] // can even add this to avoid confusion for new code consuming this library
  public static int GetNumLetters(this string This) => This.NumLetters;
  [EditorBrowsable(EditorBrowsableState.Never)] // can even add this to avoid confusion for new code consuming this library
  [OverloadResolutionPriority(-1)] // do we need this to tell new code to go with StringExt's method, 
  // or can we as a rule "de-prioritize" extension method below extension type's method automatically?
  public static int GetNumDigits(this string This) => ((StringExt)This).GetNumDigits();
}
```

**Reply by CyrusNajmabadi** [reply]
> If they want to offer extension types but also keep the extension methods for compatibility, that's also possible:

Yes. Again, as stated, you can always use the old form. The question is can we have an approach that doesn't force people to use the old form. 

No one is saying we can't use the approach you mentioned. We are saying it has unfortunate drawbacks for a subset of users. So the question is: is there a design that has no drawbacks for either groups.

**Reply by TahirAhmadov** [reply]
How about a source generator, potentially even an MS provided one, which generates that boilerplate? Or even a compiler feature if that's so important?
```cs
[GenerateExtensionMethods("StringExtensions")] // SG approach
extension StringExt for string : IFoo, IBar as StringExtensions // or language feature
{
  [GenerateExtensionMethod(Prefix = "Get")]
  public int NumLetters => this.Where(c=>char.IsLetter(c)).Count();
  public int GetNumDigits() => this.Where(c=>char.IsDigit(c)).Count();
}
// the StringExtensions static class gets generated either by the SG or auto-emitted by the compiler, as needed
```

**Reply by CyrusNajmabadi** [reply]
What issue would that be solving?

**Reply by TahirAhmadov** [reply]
It allows library authors to start using new extensions, while keeping the resulting DLL binary compatible for their customers who wish to continue using existing extension methods, with minimum extra effort on their part, while also significantly reducing the design and implementation workload for C# itself. Win win win.

**Reply by CyrusNajmabadi** [reply]
We'd have to design and implement the SG that then perfectly kept back compat.

**Reply by TahirAhmadov** [reply]
Yes, and that would be way cheaper than trying to figure out how to emit extensions in a way to be binary compatible with extension methods, while thinking ahead to future changes to support interfaces, and planning potential breaking changes around "typeness" of extensions, etc etc

**Reply by CyrusNajmabadi** [reply]
> Yes, and that would be way cheaper than trying to figure out how to emit extensions in a way to be binary compatible with extension methods,

How is it cheaper?

**Reply by TahirAhmadov** [reply]
Not cheaper - it's free to you because I'll write it for you. 🤣 not even joking

**Reply by CyrusNajmabadi** [reply]
How will you write it to ensure perfect back compat?  How are you even going to identify the differences?

**Reply by TahirAhmadov** [reply]
I hinted at it earlier, I imagine some rules, options controlled by attributes, etc. Keep in mind that if all else fails, they can always just write it out manually, but I'm certain the generator can do 99%

**Reply by CyrusNajmabadi** [reply]
> Keep in mind that if all else fails, they can always just write it out manually

Then they don't have a safe path forward.  The idea is to make it 100% safe so there is no compat concern.  If it can potentially break they're not going to risk it.

**Reply by CyrusNajmabadi** [reply]
This isn't a minor concern.  There are literally 20 years of work built in top of extension methods, with all their capabilities and flexibility.

**Reply by TahirAhmadov** [reply]
The fact that I am convincing a member of LDT to use a generator is hilarious 🤣

I really don't see that big of an issue here. They can quickly inspect the generated extension methods and ensure that the signatures match up. Keep in mind that the code fixer will be doing the conversion; the chances of typos etc are non existent. 
Between clearly defined rules, an appropriate level of customization, and the ability to manually write the extension methods to redirect to the new extensions for the edge cases, it's clear to me that this isn't a serious problem to justify degrading the rest of the feature.

**Reply by CyrusNajmabadi** [reply]
> it's clear to me that this isn't a serious problem

this is a primary concern.  it may not be to you, but that doesn't change how important it is for the ecosystem as a whole.

**Reply by TahirAhmadov** [reply]
To clarify, I'm not saying it should be ignored because it's not a problem to me, or that I don't maintain such libraries with extension methods. My point is only that, approaching this with full understanding of the challenge, I think the available solution is satisfactory.

**Comment by alrz** [comment]
> We'd then further be concerned that this would block the BCL from using the new form in conjunction with the old form,

To that end, would it make sense to allow the old form `this` in extensions instead of `for` on members? Essentially https://github.com/dotnet/csharplang/discussions/7515

> due to concerns about the differing inference methodology, as well as other behavioral differences between the forms.

However, I didn't understand how switching between the forms is not a breaking change if there are behavioral differences.

**Reply by CyrusNajmabadi** [reply]
> To that end, would it make sense to allow the old form this in extensions instead of for on members?

Quite possibly.  We're dsicussing exactly that.

> However, I didn't understand how switching between the forms is not a breaking change if there are behavioral differences.

The main question is: will there be behavioral differences?  Or can we ensure users can always use the new form in a way that ensures *no* behavioral differences.
